Misc changes to the size_report script

dcpleung opened the pull request:
Misc changes to the size_report script to make it a bit clear on where ROM/RAM are used. See individual commits for details.

dcpleung committed:
scripts: size_report: fix pylint docstrings and import issues

This fixes the issues on import statements identified by pylint.
Also add docstrings for method and classes. Most of them are
already there as comments, so simply move them as docstrings.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
scripts: size_report: recognize C++ file extensions

This adds some common C++ file extensions so the script
can recognize those as source file, and display them in
different color.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
scripts: size_report: keep full path until inserting into tree

This changes how paths are stored in intermediate structures
so that full paths are stored. This makes it more consistent
with those structures to avoid an issue where some paths are
full paths, some are relateive to ZEPHYR_BASE.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
scripts: size_report: alter how nodes are grouped together

This changes how the tree is generated. Symbols with no paths
are now grouped together instead of scattering throughout
the tree. If paths in ELF file are not all under ZEPHYR_BASE,
it will create additional node groups as 2nd level. This is
useful when not all source files are under ZEPHYR_BASE, and
provides a better indications of where they are.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
scripts: size_report: add tree node for hidden symbols

The list of symbols only contain visible symbols that can be
obtained through ELF file. There are code and data where
the toolchain never emits symbols for them and thus are not
visible in the list. So add a "(hidden)" node to the tree
to show they are there.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

carlescufi reviewed (approved):
Tested this locally, looks much better. Thanks!

carlescufi commented:
I think we could improve further by using `WEST_TOPDIR` instead of the full path for anything that is outside `ZEPHYR_BASE`, but we can do that in a second pass.

carlescufi reviewed (changes requested):
Actually, there is a regression here:

Before:

```
├── modules                                                                                      1378 8.19%
│   ├── debug                                                                                     100 0.59%
│   │   └── segger                                                                                100 0.59%
│   │       └── SEGGER                                                                            100 0.59%
│   │           └── SEGGER_RTT.c                                                                  100 0.59%
│   │               ├── SEGGER_RTT_Init                                                             4 0.02%
│   │               └── _DoInit                                                                    96 0.57%
```

now:
```
├── /                                                                                            1354  8.01%
│   └── home                                                                                     1354  8.01%
│       └── carles                                                                               1354  8.01%
│           └── src                                                                              1354  8.01%
│               └── zephyr                                                                       1354  8.01%
│                   └── modules                                                                  1354  8.01%
│                       ├── debug                                                                 100  0.59%
│                       │   └── segger                                                            100  0.59%
│                       │       └── SEGGER                                                        100  0.59%
│                       │           └── SEGGER_RTT.c                                              100  0.59%
│                       │               ├── SEGGER_RTT_Init                                         4  0.02%
│                       │               └── _DoInit                                                96  0.57%
```

dcpleung committed:
scripts: size_report: add tree node for WORKSPACE

Add a tree node to group files under WORKSPACE so that
they won't be shown with full path. The WORKSPACE is
usually the same as WEST_TOPDIR unless ZEPHYR_WORKSPACE
is defined during build.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Fixed by adding a `WEST_TOPDIR` node to group all files inside west workspace under this node.

nashif commented:
```
Path                                                                                             Size    %
==============================================================================================================
Symbols                                                                                          9618 100.00%
├── (hidden)                                                                                        0  0.00%
├── (no paths)                                                                                   4752 49.41%
│   ├── _kernel                                                                                    44  0.46%
│   ├── _sw_isr_table                                                                             688  7.15%
│   ├── dynamic_regions.0                                                                          12  0.12%
│   ├── g_xtal0Freq                                                                                 4  0.04%
│   ├── g_xtal32Freq                                                                                4  0.04%
│   ├── k_sys_work_q                                                                              192  2.00%
│   ├── my_stack_area                                                                             544  5.66%
│   ├── my_stack_area_0                                                                           544  5.66%
│   ├── my_thread                                                                                 160  1.66%
│   ├── my_thread_0                                                                               160  1.66%
│   ├── z_idle_threads                                                                            160  1.66%
│   ├── z_interrupt_stacks                                                                       2080 21.63%
│   └── z_main_thread                                                                             160  1.66%
├── WEST_TOPDIR                                                                                     8  0.08%
│   └── modules                                                                                     8  0.08%
│       └── hal                                                                                     8  0.08%
│           └── nxp                                                                                 8  0.08%
│               └── mcux                                                                            8  0.08%
│                   └── devices                                                                     8  0.08%
│                       └── MK64F12                                                                 8  0.08%
│                           └── fsl_clock.c                                                         8  0.08%
│                               ├── s_fastIrcFreq                                                   4  0.04%
│                               └── s_slowIrcFreq                                                   4  0.04%
└── ZEPHYR_BASE                                                                                  4858 50.51%
    ├── arch                                                                                        1  0.01%
    │   └── arm                                                                                     1  0.01%
    │       └── core                                                                                1  0.01%
    │           └── aarch32                                                                         1  0.01%
    │               └── mpu                                                                         1  0.01%
    │                   └── nxp_mpu.c                                                               1  0.01%
    │                       └── static_regions_num                                                  1  0.01%
    ├── drivers                                                                                   156  1.62%
    │   ├── clock_control                                                                           8  0.08%
    │   │   ├── clock_control_mcux_mcg.c                                                            4  0.04%
    │   │   │   └── __devstate_dts_ord_35                                                           4  0.04%
    │   │   └── clock_control_mcux_sim.c                                                            4  0.04%
    │   │       └── __devstate_dts_ord_8                                                            4  0.04%
    │   ├── console                                                                                 4  0.04%
    │   │   └── uart_console.c                                                                      4  0.04%
    │   │       └── uart_console_dev                                                                4  0.04%
```

tejlmand commented on cmake/reports/CMakeLists.txt:
```diff
@@ -12,6 +12,7 @@ foreach(report ram_report rom_report)
     -k ${ZEPHYR_BINARY_DIR}/${KERNEL_ELF_NAME}
     -z ${ZEPHYR_BASE}
     -o ${CMAKE_BINARY_DIR}
+    $<$<BOOL:${WEST_TOPDIR}>:--westtopdir=${WEST_TOPDIR}>
```
Should this be a made a bit more generic ?

size_report itself doesn't have any dependencies on west, and it could probably be useful for users not using west to still specify a workspace dir, as example using `-DZEPHYR_WORKSPACE=<workspace path>`.

Also, we already at this stage know whether or not WEST_TOPDIR is defined, so there is little value in using a generator expression here.
Generator expressions are very powerful when it comes to avoid configure time process order dependencies, as example allowing target properties to be set late in cmake.
But they are harder to read for most users.

So I propose to change into:
```
...
set(report_depth 99)
if(DEFINED ZEPHYR_WORKSPACE)
  set(workspace_arg "--workspace=${ZEPHYR_WORKSPACE}")
elseif(DEFINED WEST_TOPDIR)
  set(workspace_arg "--workspace=${WEST_TOPDIR}")
endif()

foreach(report ram_report rom_report)
  add_custom_target(
    ${report}
    ${PYTHON_EXECUTABLE}
    ${ZEPHYR_BASE}/scripts/footprint/size_report
    -k ${ZEPHYR_BINARY_DIR}/${KERNEL_ELF_NAME}
    -z ${ZEPHYR_BASE}
    -o ${CMAKE_BINARY_DIR}
    ${workspace_arg}
   ....
```
which also follows the pattern for other arguments, such as `flag_for_${report}` and `report_depth`.

tejlmand commented on scripts/footprint/size_report:
```diff
@@ -691,6 +713,8 @@ def parse_args():
                         help="Zephyr base path")
     parser.add_argument("-o", "--output", required=True,
                         help="Output path")
+    parser.add_argument("-w", "--westtopdir", default=None,
+                        help="West path (WEST_TOPDIR)")
```
This size report script itself is independent of `west`, so I propose to make it a generic `--workspace` argument.

This would indicate that it is also useful for non-west users, but where code is still located in a common folder / workspace.

```suggestion
    parser.add_argument("-w", "--workspace", default=None,
                        help="Workspace path")
```

nashif replied:
makes sense

tejlmand reviewed (changes requested):
Looks good, but I propose to make this a generic `workspace` argument.

tejlmand reviewed (approved):
Looks good.
Thanks for the applied changes :+1:

carlescufi commented:
Looks great, thank you!
